README
----------------------------------------------------------------
- Wrathion - A GPU accelerated document password recovery tool -
- Authors: Jan Schmied, Radek Hranicky, Vladimir Vesely, Peter Gazdik -
----------------------------------------------------------------

Wrathion was successfully compiled and tested under Ubuntu 14.04 LTS and Windows 8.1
CPU-only version should work on all common Intel CPUs.
GPU-accelerated version was DEVELOPED PRIMARILY for AMD cards, but also works with NVIDIA cards.

Table of contents:
I-a. PREREQUISITES - Ubuntu 14.04 LTS (Detailed "how-to cookbook")
I-b. PREREQUISITES - Other GNU/Linux OS
I-c. PREREQUISITES - Windows 8.1 or other
II. COMPILATION
III. USAGE

The easiest way of getting Wrathion operational is to use a PC with AMD GPU,
fresh installation of Ubuntu 14.04 LTS, and following steps I-a., II., and III.

#######################################################################
# I-a. PREREQUISITIES - UBUNTU 14.04 LTS (Detailed "how-to cookbook") #
#######################################################################

In Ubuntu 14.04 LTS ALL prerequisities should be obtainable in the form of packages
which you install by "sudo apt-get install PACKAGENAME" or by using Synaptic package
manager - type: "sudo apt-get install synaptic", "sudo apt-get update", and run synaptic.

1. Install packages for Wrathion core:
build-essential
cmake
ubuntu-drivers-common
doxygen (optional)

2. Install packages required by ZIP module:
zlib1g
zlib1g-dev

3. Install packages required by PDF module:
libpoppler44
libpoppler-dev
libpoppler-private-dev
libpoppler-glib8
poppler-data

4. IMPORTANT: Execute the following command (solves error in poppler-goo):
sudo cp /usr/include/poppler/poppler-config.h /usr/include

(5a). IF you use AMD GPU:
# FOR AMD CARDS: Radeon(TM) R9 Fury, R9 300, R9 200, R7 300, HD 7000, HD 6000, and HD 5000 Series
# WAS TESTED ON: R9 290X and R9 Fury X
Install the following packages (Proprietary driver, etc.):
fglrx-updates
fglrx-updates-core
fglrx-updates-dev
fglrx-amdcccle-updates
ocl-icd-libopencl1
ocl-icd-opencl-dev
opencl-headers

(5b). IF you use NVIDIA GPU, you need following packages
# SHOULD WORK ON ALL NVIDIA cards that support OpenCL: https://www.khronos.org/conformance/adopters/conformant-products
# WAS TESTED ON: GeForce 660Ti
Install the following packages (Proprietary driver, etc.):
xserver-xorg-video-nouveau
nvidia-331-updates
nvidia-331-updates-uvm
nvidia-331-updates-dev
nvidia-opencl-icd-311-updates
nvidia-prime
nvidia-settings
ocl-icd-dev
opencl-headers
(NOTE: Newer drivers MAY also work with your GPU. Drivers denoted above were tested with GeForce 660Ti)

-----------------------------------------------------------------------------------------
 IF YOU PROCEEDED THROUGH STEPS 1-5, YOU SHOULD BE NOW ABLE TO COMPILE AND RUN WRATHION!
-----------------------------------------------------------------------------------------

* Tested again on 2016-01-21 on fresh installation of Ubuntu 14.04 LTS with the following hardware:
* ASUS ROG RAMPAGE V EXTREME
* Intel Core i7-5930K
* 32 GB RAM DDR4 Kingston
* 2x AMD R9 Fury X

############################################
# I-b. PREREQUISITIES - Other GNU/Linux OS #
############################################

Depending on your Linux distribution, you need to install dependencies in the similar
way as shown in section I-a. The type of OpenCL library and OpenCL kernel compilation
are GPU-dependent, thus, we highly recommend to use proprietary drivers from your
GPU manufacturer. Using non-proprietary drivers may cause troubles with kernel compilation
which may limit Wrathion to CPU-only version.

IF your distribution does not support AMD Catalyst (fglrx) / NVIDIA drivers in the form
of package, you may need to download and install them manually from your GPU manufacturer's
website.

- Poppler library may be obtained from: http://poppler.freedesktop.org/ (tested with v0.39)
- ZLib library may be obtained from: http://www.zlib.net/
- OpenCL headers may be obtained from Khronos OpenCL Registry: https://www.khronos.org/registry/cl/


################################################
# I-c. PREREQUISITIES - Windows 8.1 (or other) #
################################################

NOTE: Compiling Wrathion of Windows is a bit more tricky than on GNU/Linux.
      With steps below, we compiled and tested Wrathion on Windows 8.1 N:

1. WHAT YOU NEED:
Proprietary AMD Catalyst / NVIDIA graphics driver
MinGW with GCC compiler: http://www.mingw.org/
(We recommend to use automated mingw-get-setup.exe obtainable from https://sourceforge.net/projects/mingw/files/latest/download)
MSYS (Minimalist GNU for Windows): http://www.mingw.org/wiki/msys
POSIX Threads for Windows: http://sourceforge.net/projects/pthreads4w/
LibPoppler 0.39 or higher - we recommend to use precompiled binaries: http://sourceforge.net/projects/poppler-win32/
OpenCL headers: https://www.khronos.org/registry/cl/
For AMD you MAY NEED AMD APP SDK: http://developer.amd.com/tools-and-sdks/opencl-zone/amd-accelerated-parallel-processing-app-sdk/
cmake and make (for Windows) - we recommend installing by mingw-get-setup.exe:
https://github.com/cpluspluscom/ChessPlusPlus/wiki/MinGW-Build-Tutorial

2. ENSURE that you have all dependencies in the correct (bin, lib, include) directories:
- .h and .hpp files go to "include" directory
- .dll files go to "bin" directory
- .a and .lib files go to "lib" directory

FOR EXAMPLE:
C:\MinGW\x86_64-w64-mingw32\include\pthread.h
C:\MinGW\x86_64-w64-mingw32\include\poppler\*
C:\MinGW\x86_64-w64-mingw32\include\poppler\cpp\*
C:\MinGW\x86_64-w64-mingw32\include\poppler\CL\cl.h
C:\MinGW\x86_64-w64-mingw32\include\poppler\CL\cl.hpp
...
C:\MinGW\x86_64-w64-mingw32\bin\libpoppler.dll
C:\MinGW\x86_64-w64-mingw32\bin\zlib1.dll
C:\MinGW\x86_64-w64-mingw32\bin\OpenCL.dll
...
C:\MinGW\x86_64-w64-mingw32\lib\libpthread.a
C:\MinGW\x86_64-w64-mingw32\lib\libpthread.dll.a
C:\MinGW\x86_64-w64-mingw32\lib\libpoppler.a
C:\MinGW\x86_64-w64-mingw32\lib\libpoppler.dll.a
C:\MinGW\x86_64-w64-mingw32\lib\OpenCL.lib
...

3. ENSURE that the PATH to make, cmake, etc. is set properly, e.g.:
SET PATH=c:\MinGW\bin;%PATH%

4. RUN MSYS (not MinGW) CONSOLE - e.g.: c:\msys\1.0\msys.bat
- You will use UNIX commands from now!
- Ensure that MinGW directory is mounted properly, if not - do it (e.g.: mount c:/mingw /mingw)
- Go to the Wrathion source directory
- Proceed to section II.

###################
# II. COMPILATION #
###################

Wrathion framework can be compiled using CMake (version 2.8 and above).
On Windows (tested on MinGW MSYS).

cmake -G"MSYS Makefiles"
make

On Linux (tested on Ubuntu 13.10 and 14.04)

cmake -G"Unix Makefiles"
make

** SOLUTION OF POSSIBLE PROBLEMS **
- IF you see any undefined reference to 'pthread_...', re-run the cmake and make again
  (Radek: 2016-01-21: SHOULD BE FIXED NOW, should not occur again.)
- IF there was already cmake command executed before, you can erase the configuration by
deleting CMakeCache.txt and CMakeFiles in all Wrathion subdirectories.
- IF you encounter any other errors, please check if ALL prerequisities are satisfied (section I.).

All binary files are placed into 'bin' directory. OpenCL kernels are copyed from
its directories inside mudules into 'bin/kernels' dorectory. This happens when
running cmake command. If you delete contents of 'bin/kernels' directory its
content is not restored by running make. Re-run cmake or restore contents
manualyl.

If youhave doxygen installed, you can generate a documentation of Wrathion by typing:
doxygen doxygen.config

# HOW TO TEST OPENCL AVAILABILITY
After the compilation (see below), you may also check the correctness of your build by typing
./wrathion -s
Or
wrathion.exe -s
(in your bin/ directory):
The output should contain all your OpenCL-compatible GPUs installed in your computer.
Another useful command for OpenCL details about your GPUs is:
clinfo

##############
# III. USAGE #
##############

After the compilation, there should be a binary executable file "wrathion" or "wrathion.exe" in bin/ directory.
It supports the following arguments:
    -h - prints this help
    -f - input file
    --modules -l - list loaded modules
    --devices -s - list platforms and devices
    --cpu-cracker -c - prefer CPU cracker over GPU generator
    --map -d - devices to use <platform>:<device>[:<GWS>][,<platform>:<device>[:<GWS>],...]
    --chars -p - chars for creating passwords (default: abcdefghijklmnopqrstuvwxyz)
    -u (alternative of --chars) - file with unicode characters in hex form
    -m - maximum length of password (default: 10)
    --dict=file, -r - dictionary for dictinary attack
    --threads=NUMTHREADS, -t - number of threads for CPU Cracking
    -v - verbose mode (more information is displayed)

Markov attack
    -S, --statistics        file with statistics
    -T, --thresholds=glob[:pos]
                            number of characters per position
          - glob - global value for every position in password (default 95)
          - pos - positional comma-separated values(overwrites global value)
    -L, --length=min:max    length of password (default 1:50)
    -M, --mask              mask
    -X, --model             type of Markov model:
          - classic - First-order Markov model (default)
          - layered - Layered Markov model
    -I - return password on given index (only for experiments)
    -C, --cpu-generator - prefer CPU generator over GPU generator (works only
                          with a Markov generator)


Example:
./wrathion -vd 0:0 -f testfiles/orig/aes256.zip -S stats/rockyou.wstat -L 1:10
           -X layered -T 26 -M ?u?l?l?l?l?l?d?d?d?d (on Ubuntu 14.04)
Or:
wrathion.exe -vd 0:0 -f testfiles/orig/aes256.zip -S stats/rockyou.wstat -L 1:10
           -X layered -T 26 -M ?u?l?l?l?l?l?d?d?d?d  (on Windows 8.1)

Detailed decription of Markov attack is in file MANUAL (only in Slovak)

The directory /bin/testfiles contains examples of various encrypted documents.
The directory /bin/uchars contains examples of unicode alphabet files.
The directory /bin/stats contains various files with statistics for First-order and Layered Markov model.

Unicode alphabet files contain the definition of characters in Unicode U+ hex notation (e.g. "U+0042"). The "#" symbol can be used for comments. Any character sequence beginning with "#" symbol and ending with an end of the line, is ignored.

Wrathion is also able to save its generator state and resume the work later. The generator state is saved automatically in [filename].passgen file after receiving SIGINT signal. If launched again with the same parameters, Wrathion will continue its previous job from the saved state. If you wish to re-run the whole job from the beginning, just delete or rename the .passgen file.

------------------------------------------
 Radek Hranicky - Last update: 2016-01-21
------------------------------------------


CHANGES
#################################################
# Zmenené súbory
#################################################

main.cpp - pridané nové parametre, inštancovanie Markovského generátora namiesto
brute-force generátora

core/GPUCracker.cpp - upravené riadenie generovania a overovania hesla na
GPU pre potreby nového generátora (pôvodný brute-force generátor nie je ďalej
podporovaný, sú potrebné zmeny v jeho implementácii, nakoľko sa veľa vecí spojené
s generovaním vykonávalo v samotnej triede GPUCracker). Odstránené chyby
spojené s nastavovaním zlej veľkosti poľa obsahujúce heslá v GPU. Slovníkový
generátor je stále podporovaný.

core/PassGen.cpp, include/PassGen.h - v triede PassGen pridaná metóda
'nextKernelStep' kvôli novému generátoru (ako vzorové rozhranie pre dotatočnú
úpravu pôvodneho generátora pre zachovanie kompatibility - navrhujem však
jeho kompletné prepracovanie na nový spôsob generovania, nakoľko sa ukázala
pôvodná implementácia neefektívna, viď dokumentácia),
pridaný copy konštruktor pre vytvorenie nového generátora, pravdepodobne
opravené nejaké chyby.

CMakeLists.txt, core/CMakeLists.txt - linkovanie dll knižnice pod Windows,
ktorá obsahuje funkcie pre prevod hodnôt v 'network byte order' do
hostiteľskej endianity.

#################################################
# Pridané súbory
#################################################

include/MarkovPassGen.h, core/MarkovPassGen.cpp - implementácia nového
generátora podporujúca CPU aj GPU generovanie implementované v jedinej triede
'MarkovPassGen'

include/Mask.h, core/Mask.cpp - trieda pre podporu maskovania v Markovom
generátora, parsuje a ukladá masku v bitovom poli. Pre definovanie novej
masky sa využívajú lambda funkcie v hlavičkovom súbore, viď implememtáciu.

core/kernels/markov_passgen.cl - zdrojový kód GPU generátora

CHANGES - tento súbor

MANUAL - súbor obsahujúci detailnejší popis použitia Markovského generátora


main.cpp
/* 
 * Copyright (C) 2014 Jan Schmied, Radek Hranicky
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy 
 * of this software and associated documentation files (the "Software"), to deal 
 * in the Software without restriction, including without limitation the rights 
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell 
 * copies of the Software, and to permit persons to whom the Software is 
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in 
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE 
 * SOFTWARE.
 * 
 */

#include <cstdlib>
#include <getopt.h>
#include <iostream>
#include <vector>
//#include <ctime>
#include <sys/time.h>

#include "FormatPool.h"

#ifndef WRATHION_DLL_MODULES //compile modules directly 
#include "PDFFormat.h"
#include "ZIPFormat.h"
#include "DOCFormat.h"
#endif

#include "GPUCracker.h"
#include "CrackerRunner.h"
#include "Module.h"
#include <csignal>
#include <ctime>
#include <cstdint>
#include "UnicodeParser.h"
#include <MarkovPassGen.h>

#ifdef WRATHION_MPI
#include <mpi.h>
#endif

/*
 * Maximum size of possible UTF8 characters per each UTF32 character
 * (For the purpose of UnicodePassGen)
 */
#define UTF8_CHAR_MAXSIZE 4

#define CLOCKS_PER_MSEC (CLOCKS_PER_SEC/1000)
#define CLOCKS_PER_USEC (CLOCKS_PER_SEC/1000000)

using namespace std;

const string help = "wrathion [OPTIONS]\n"
"    -h - prints this help\n"
"    -f - input file\n"
"    --modules -l - list loaded modules\n"
"    --devices -s - list platforms and devices\n"
"    --cpu-cracker -c - prefer CPU cracker over GPU generator\n"
"    --map -d - devices to use <platform>:<device>[:<GWS>][,<platform>:<device>[:<GWS>],...]\n"
"    --chars -p - chars for creating passwords (default: abcdefghijklmnopqrstuvwxyz)\n"
"    -u (alternative of --chars) - file with unicode characters in hex form\n"
"    -m - maximum length of password (default: 10)\n"
"    --dict=file, -r - dictionary for dictinary attack\n"

#ifdef WRATHION_MPI
"    --mpi - run in MPI mode\n"
#endif
"    --threads=NUMTHREADS, -t - number of threads for CPU Cracking\n"
"    -v - verbose mode (more information is displayed)\n"
"\n"
"Markov attack\n"
"    -S, --statistics        file with statistics\n"
"    -T, --thresholds=glob[:pos]\n"
"                            number of characters per position\n"
"          - glob - global value for every position in password (default 95)\n"
"          - pos - positional comma-separated values(overwrites global value)\n"
"    -L, --length=min:max    length of password (default 1:50)\n"
"    -M, --mask              mask\n"
"    -X, --model             type of Markov model:\n"
"          - classic - First-order Markov model (default)\n"
"          - layered - Layered Markov model\n"
"    -I - return password on given index\n"
"    -C, --cpu-generator - prefer CPU generator over GPU generator (works only\n"
"                          with a Markov generator)\n"
"\n"
"Mask syntax: (same syntax as Hashcat and John the Ripper)\n"
"    ?l - lower-case letters (abcdefghijklmnopqrstuvwxyz)\n"
"    ?u - upper-case letters (ABCDEFGHIJKLMNOPQRSTUVWXYZ)\n"
"    ?d - digits (0123456789)\n"
"    ?s - punctuation and symbols (<<space>>!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~)\n"
"    ?a - all previous (?l?u?d?s)\n";

struct opts : MarkovPassGen::Options {
    opts():
        help(false),
        show_modules(false),
        show_devices(false),
        prefer_cpu_cracker(false),
        max_pass_len(10),
        stdin_mode(true),
#ifdef WRATHION_MPI       
        mpi(false),
#endif
        verbose(false)
    {}
    bool help;
    string input_file;
    bool show_modules;
    bool show_devices;
    bool prefer_cpu_cracker;
    bool prefer_cpu_generator = false;
    string devices_mapping;
    char chars[256] = {0};
    UnicodeParser unicodeParser;
    int max_pass_len = 10;
    int threads = 0;
    string dict;
    string unicode_file;
    bool stdin_mode;
#ifdef WRATHION_MPI
    bool mpi;
#endif
    bool verbose;
    uint64_t index = UINT64_MAX;
};

bool stop = false;

void sigint_handler (int param)
{
  stop = true;
  cout << "CTRL+C catched" << endl;
}


inline long long int getMilliSecs() { 
    timeval t;    
    gettimeofday(&t, NULL);
    return (t.tv_sec) * 1000 + (t.tv_usec) / 1000;
}


/*
 * 
 */
int main(int argc, char** argv) {
    long long int starttime, elapsedtime;
    if (argc < 2){
        cout << help;
        return 0;
    }
    int opt,opt_index;
    opts o;
    static struct option long_options[] =
             {
               {"modules", no_argument, 0, 'l'},
               {"devices", no_argument, 0, 's'},
               {"dict",    required_argument, 0, 'r'},
               {"threads",  required_argument, 0, 't'},
               {"cpu-cracker",  no_argument, 0, 'c'},
               {"cpu-generator",  no_argument, 0, 'C'},
               {"map",  required_argument, 0, 'd'},
               {"chars",    required_argument, 0, 'p'},
#ifdef WRATHION_MPI       
               {"mpi", no_argument, 0, 'z'},
#endif
							 // Markov attack
							 {"statistics", required_argument, 0, 'S'},
							 {"thresholds", required_argument, 0, 'T'},
							 {"length", required_argument, 0, 'L'},
							 {"model", required_argument, 0, 'X'},
							 {"mask", required_argument, 0, 'M'},
               {0, 0, 0, 0}
             };
    while ((opt = getopt_long(argc, argv, "hf:lscd:p:u:r:t:vm:S:T:L:M:X:I:C", long_options,&opt_index)) != -1){
        switch(opt){
        	  case 'S':
        	  	o.stat_file = optarg;
        	  	break;
        	  case 'T':
        	  	o.thresholds = optarg;
        	  	break;
        	  case 'L':
        	  	o.length = optarg;
        	  	break;
        	  case 'X':
        	  	o.model = optarg;
        	  	break;
        	  case 'M':
        	  	o.mask = optarg;
        	  	break;
        	  case 'I':
        	    o.index = strtoull(optarg, nullptr, 0);
        	    break;
        	  case 'C':
        	    o.prefer_cpu_generator = true;
        	    break;
            case 'h':
                o.help = true; break;
            case 'f':
                o.input_file.assign(optarg); break;
            case 'l':
                o.show_modules = true; break;
            case 's':
                o.show_devices = true; break;
            case 'c':
                o.prefer_cpu_cracker = true; break;
            case 'd':
                o.devices_mapping.assign(optarg); break;
            case 'p':
                ::strcpy(o.chars,optarg); break;
            case 'u':
                o.unicode_file.assign(optarg); break;
            case 'm':
                o.max_pass_len = atoi(optarg); break;
            case 'r':
                o.dict.assign(optarg); break;
            case 't':
                o.threads = atoi(optarg); break;
            case 'v':
                o.verbose = true; break;
#ifdef WRATHION_MPI  
            case 'z':
                o.mpi = true; break;
#endif                  
        }
    }
    
    if (o.chars[0] != 0 && !o.unicode_file.empty()) {
        cout << "Please specify either -p or -u, not both." << endl;
        return 0;
    }
    
    signal(SIGINT, sigint_handler);
    
    if(o.chars[0] == 0){
        ::strcpy(o.chars,"abcdefghijklmnopqrstuvwxyz");
    }
    
    FormatPool pool;
    if (o.verbose){
        pool.setVerbose();
    }
    
#ifdef WRATHION_DLL_MODULES //load modules form shared libs 
    vector<Module*> modules = Module::loadModules();
    for(int i = 0;i<modules.size();i++){
        pool.addFormat(modules[i]->getFileFormat());
    }
#else
    pool.addFormat(new PDFFormat());
    pool.addFormat(new ZIPFormat());
    pool.addFormat(new DOCFormat());
#endif   
    bool run = false;
    FileFormat* format = NULL;
    CrackerFactory *crackerFactory = NULL;
    PassGen *passgen = NULL;
    CrackerRunner runner;
    
    if (o.verbose){
        runner.setVerbose();
    }
            
    if(o.help){
        cout << help;
        return 0;
    }
    if(o.show_modules){
        Module::printLoadedModules(&cout);
        return 0;
    }
    if(o.show_devices){
        cout << GPUCracker::availableDevices();
        GPUCracker::destroyOpenCL();
        return 0;
    }
    if(!o.stat_file.empty() && o.index != UINT64_MAX)
    {
      MarkovPassGen passgen {o};
      string pass = passgen.getPassword(o.index);
      cout << pass << endl;
      return 0;
    }
    if (o.dict.empty() && o.unicode_file.empty() &&  o.stat_file.empty())
    {
      cout << help;
      return 0;
    }
    if(!o.input_file.empty()){
        format = pool.getFileFormat(o.input_file); // Detection of file format
        if(format == NULL){
            cout << "No suitable format found" << endl;
        }else{
            if (!format->isEncrypted()) {
                cout << "This file is NOT encrypted." << endl;
                return 0;
            }
            if(o.prefer_cpu_cracker){
                cout << "Trying to get CPU cracker" << endl;
                crackerFactory = format->getCPUCracker();
                if(crackerFactory == NULL){
                    cout << "CPU cracker not found" << endl << "Trying to get GPU cracker" << endl;
                    crackerFactory = format->getGPUCracker();
                    if(crackerFactory == NULL){
                        cout << "GPU cracker not found" << endl << "No suitable cracker found" << endl << "Bye" << endl;
                    }else{
                        cout << "Using GPU cracker" << endl;
                        run = true;
                    }
                }else{
                    cout << "Using CPU cracker" << endl;
                    run = true;
                }
            }else{
                cout << "Trying to get GPU cracker" << endl;
                crackerFactory = format->getGPUCracker();
                if(crackerFactory == NULL){
                    cout << "GPU cracker not found" << endl << "Trying to get CPU cracker" << endl;
                    crackerFactory = format->getCPUCracker();
                    if(crackerFactory == NULL){
                        cout << "CPU cracker not found" << endl << "No suitable cracker found" << endl << "Bye" << endl;
                    }else{
                        cout << "Using CPU cracker" << endl;
                        run = true;
                    }
                }else{
                    cout << "Using GPU cracker" << endl;
                    run = true;
                }
            }
        }
    }else{
        cout << "No input file" << endl;
        return 0;
    }
    
    // In case of GPU cracker we need additional 1 uchar point for current password length
    if (run && crackerFactory->isGPU()) {
        o.max_pass_len++;
    }
    
    if(run){
        if (!o.unicode_file.empty()) {
            int chars_count;
            if (!o.unicodeParser.processFile(o.unicode_file, &chars_count)) {
                return 1;
            }
            o.unicodeParser.getCharsPtr();
            passgen = new UnicodePassGen(o.unicodeParser.getCharsPtr(), chars_count, o.max_pass_len*UTF8_CHAR_MAXSIZE, o.max_pass_len);
        } else if (!o.dict.empty()){
            passgen = new DictionaryPassGen(o.dict);
        } else {
            passgen = new MarkovPassGen(o, o.prefer_cpu_generator);
        }
        passgen->loadState(o.input_file+".passgen");
        
        if(o.devices_mapping != ""){
            // In the case we have any manually mapped device
            std::vector<DeviceConfig> confs = CrackerRunner::decodeConfig(o.devices_mapping);
            if(confs.size() > 0){
                runner.setDeviceConfig(confs);
            }
        }
        
        runner.setCrackerFactory(crackerFactory);  // set cracker factory
        runner.setPassGen(passgen);                // set password generator
        if(o.threads > 0){
            runner.setNumThreads(o.threads);
        }
        // Time measurement
        if (o.verbose){
            starttime = getMilliSecs();
        }
        //
        runner.start();
        cout << "Running " << runner.getNumThreads() << " threads" << endl;
        CrackerRunner::sleep(1000);
        while(runner.someRunning() && !stop){
            cout << "\r";
            vector<uint64_t> *speeds = runner.getSpeeds();
            uint64_t sum = 0;
            uint32_t n = 1;
            cout << "Speed: ";
            for(vector<uint64_t>::iterator i = speeds->begin();i != speeds->end();i++){
                cout << "Thread " << n << ": " << *i << " p/s, ";
                sum += *i;
                n++;
            }
            cout << "Total: " << sum << " p/s";
            CrackerRunner::sleep(1000);
        }
        // Write elapsed time
        if (o.verbose){
            elapsedtime =  getMilliSecs() - starttime;
            cout << "Total time spent by cracking:" << endl;
            cout << (int)(elapsedtime/60000) << "m "
                 << ((double)(elapsedtime % 60000) / 1000.0) << "s "
                 << endl;
        }
        //
        cout << endl;
        if(runner.passFound()){
            cout << "Password found: '" << runner.getPassword() << "'" << endl;
        }else{
            if(stop){
                cout << "Saving generator state" << endl;
                runner.stop();
                passgen->saveState(o.input_file+".passgen");
            }else{
                cout << "No password found" << endl;
            }
        }
        CrackerRunner::sleep(1000);
        delete passgen;
        delete crackerFactory;
    }
    GPUCracker::destroyOpenCL();
    return 0;
}

